pkg/gadgets/socket-collector/tracer/tracer.go

```diff
@@ -0,0 +1,268 @@
+	statusBytes, err := hex.DecodeString(statusHex)
+	if err != nil {
+		prifixedLog(fmt.Sprintf("Failed to decode hex-string TCP status %q: %s", statusHex, err))
+		return ""
+	}
+
+	switch uint8(statusBytes[0]) {
+	case TCP_ESTABLISHED:
+		return "ESTABLISHED"
+	case TCP_SYN_SENT:
+		return "SYN_SENT"
+	case TCP_SYN_RECV:
+		return "SYN_RECV"
+	case TCP_FIN_WAIT1:
+		return "FIN_WAIT1"
+	case TCP_FIN_WAIT2:
+		return "FIN_WAIT2"
+	case TCP_TIME_WAIT:
+		return "TIME_WAIT"
+	case TCP_CLOSE:
+		return "CLOSE"
+	case TCP_CLOSE_WAIT:
+		return "CLOSE_WAIT"
+	case TCP_LAST_ACK:
+		return "LAST_ACK"
+	case TCP_LISTEN:
+		return "LISTEN"
+	case TCP_CLOSING:
+		return "CLOSING"
+	case TCP_NEW_SYN_RECV:
+		return "NEW_SYN_RECV"
+	default:
+		prifixedLog(fmt.Sprintf("Invalid TCP status: %s (%v)", statusHex, statusBytes))
+	}
+
+	return "UNKNOWN"
+}
+
+func parseUDPStatus(statusHex string) string {
+	statusBytes, err := hex.DecodeString(statusHex)
+	if err != nil {
+		prifixedLog(fmt.Sprintf("Failed to decode hex-string UDP status %q: %s", statusHex, err))
+		return ""
+	}
+
+	switch uint8(statusBytes[0]) {
+	case TCP_ESTABLISHED:
+		return "ACTIVE"
+	case TCP_CLOSE:
+		return "INACTIVE"
+	default:
+		prifixedLog(fmt.Sprintf("Unexpected UDP status: %s (%v)", statusHex, statusBytes))
+	}
+
+	return "UNKNOWN"
+}
+
+func getTCPIter() *link.Iter {
+	objs := IterTCPv4Objects{}
+	if err := LoadIterTCPv4Objects(&objs, nil); err != nil {
+		prifixedLog(fmt.Sprintf("Failed to load TCP BPF objects: %v", err))
+		return nil
+	}
+	defer objs.Close()
+
+	it, err := link.AttachIter(link.IterOptions{
+		Program: objs.DumpTcp4,
+	})
+	if err != nil {
+		prifixedLog(fmt.Sprintf("Failed to attach TCP BPF iterator: %v", err))
+		return nil
+	}
+
+	return it
+}
+
+func getUDPIter() *link.Iter {
+	objs := IterUDPv4Objects{}
+	if err := LoadIterUDPv4Objects(&objs, nil); err != nil {
+		prifixedLog(fmt.Sprintf("Failed to load UDP BPF objects: %v", err))
+		return nil
+	}
+	defer objs.Close()
+
+	it, err := link.AttachIter(link.IterOptions{
+		Program: objs.DumpUdp4,
+	})
+	if err != nil {
+		prifixedLog(fmt.Sprintf("Failed to attach UDP BPF iterator: %v", err))
+		return nil
+	}
+
+	return it
+}
+
+func RunCollector(pid uint32, filter *gadgetv1alpha1.ContainerFilter) (string, error) {
+	if err := increaseRlimit(); err != nil {
+		return "", fmt.Errorf("failed to increase memlock limit: %s", err)
+	}
+
+	// TODO: Use filters to collect only TCP or only UDP information
+	iters := []*link.Iter{}
+	iters = append(iters, getTCPIter())
+	iters = append(iters, getUDPIter())
+
+	// TCP/UDP BPF Iters read information from the current namespace thus
+	// we need to change network namespace before opering the iterator
+	netnsHandle, err := netns.GetFromPid(int(pid))
+	if err != nil {
+		return "", fmt.Errorf("failed to retrieve network namespace from PID %d: %w", pid, err)
+	}
+	defer netnsHandle.Close()
+
+	if err := netns.Set(netnsHandle); err != nil {
```

Comment:
Done. Could be useful in our case to come back to the original network namespace after using the iterators?

Reply:
Actually it's not only useful but needed. If the golang thread is let in other network namespace it could cause problems to other goroutines that will be scheduled in it. It should ideally be done by using `defer` to cover all the cases where this code could return when there is an error.

Reply:
It means that in [this example](https://github.com/vishvananda/netns#example), they switched back to the original namespace just to show that, in case of need, one can do it before returning. Otherwise, thanks to the `defer netnsHandle.Close()`, when function returns, golang thread will go automatically back to the original namespace. Is my understating correct?

Reply:
> Otherwise, thanks to the defer netnsHandle.Close(), when function returns, golang thread will go automatically back to the original namespace.

I think it's not correct. `defer netnsHandle.Close()` only closes the file descriptor of the ns handle but doesn't switch back to the original namespace. 

I think you should use something like:

```

    // Save the current network namespace
    origns, _ := netns.Get()
    defer ns.Set(origns)
    defer origns.Close()
```

Reply:
Ok, just like the example. Now also [this comment](https://github.com/kinvolk/inspektor-gadget/pull/227#discussion_r691302794) is clear. Thanks!